daemon/vpn/nordlynx/libtelio/libtelio.go

```diff
@@ -619,30 +627,88 @@ func (l *Libtelio) Public(private string) string {
 // is not caught, because receiving goroutine is not started yet. So, extra WaitGroup is used
 // to make sure this function is exited only after event receiving goroutine has started.
-func isConnected(ctx context.Context, ch <-chan state, pubKey string) <-chan bool {
+func isConnected(ctx context.Context,
+	stateCh <-chan state,
+	connParams connParameters,
+	eventsPublisher *vpn.Events) <-chan bool {
 	// we need waitgroup just to make sure goroutine has started
 	var wg sync.WaitGroup
 	wg.Add(1)
 
-	connectedC := make(chan bool)
+	connectedCh := make(chan bool)
 	go func() {
 		wg.Done() // signal that goroutine has started
-		for {
-			select {
-			case state := <-ch:
-				if state.PublicKey == pubKey &&
-					state.State == "connected" {
-					connectedC <- true
-					return
-				}
-			case <-ctx.Done():
-				connectedC <- false
-				return
-			}
-		}
+		monitorConnection(ctx, stateCh, connectedCh, connParams, eventsPublisher)
 	}()
 
 	wg.Wait() // wait until goroutine is started
 
-	return connectedC
+	return connectedCh
+}
+
+func publishConnecting(publisher *vpn.Events, server vpn.ServerData) {
+	publisher.Connected.Publish(events.DataConnect{
+		Type:                events.ConnectAttempt,
+		TargetServerIP:      server.IP.String(),
+		TargetServerCountry: server.Country,
+		TargetServerCity:    server.City,
+	})
+}
+
+func publishConnected(publisher *vpn.Events, server vpn.ServerData) {
+	publisher.Connected.Publish(events.DataConnect{
+		Type:                events.ConnectSuccess,
+		TargetServerIP:      server.IP.String(),
+		TargetServerCountry: server.Country,
+		TargetServerCity:    server.City,
+	})
+}
+
+func publishDisconnected(publisher *vpn.Events) {
+	publisher.Disconnected.Publish(events.DataDisconnect{})
+}
+
+type connParameters struct {
+	pubKey string
+	server vpn.ServerData
+}
+
+// monitorConnection awaits for incoming state changes from the states chan and publishes appropriate events. Upon
+// detecting the 'connected' state for the first time it will send true via isConnected channel and close it afterwards.
+// If goroutine is canceled before detecting 'connected', false will be sent via isConnected channel.
+func monitorConnection(
+	ctx context.Context,
+	states <-chan state,
+	isConnected chan<- bool,
+	connParameters connParameters,
+	eventsPublisher *vpn.Events) {
+	connectedPublished := false
+
+	for {
+		select {
+		case state := <-states:
+			switch state.State {
+			case "connecting":
+				publishConnecting(eventsPublisher, connParameters.server)
+			case "connected":
+				if state.PublicKey == connParameters.pubKey &&
+					state.State == "connected" {
+					if !connectedPublished {
+						isConnected <- true
+						connectedPublished = true
+						close(isConnected)
+					}
+				}
+				publishConnected(eventsPublisher, connParameters.server)
+			case "disconnected":
+				publishDisconnected(eventsPublisher)
+			}
+		case <-ctx.Done():
+			publishDisconnected(eventsPublisher)
```

Comment:
after 30 seconds this will always generate a `disconnected` event and then exit the for loop, even if it is still connected to the VPN